// aspose.barcode/examples/tests_assist.js
const barcode = require("aspose.barcode");
const fs = require("fs");
const path_to_license_file = "./resources/license/Aspose.BarCode.Nodejs.via.Java.lic";

function set_license()
{
    let license = new barcode.AsposeBarcode.assist.License();
    if (is_exists(path_to_license_file))
    {
        license.setLicense(path_to_license_file);
    }
    else
    {
        console.log("Path \"" + path_to_license_file + "\" doesn't exist\n");
    }
    let is_licensed = license.isLicensed();
    prt_mess('is_licensed => ' . is_licensed);
}

function prt_mess(mess)
{
    console.log(mess);
}

function prt_err(mess)
{
    console.log("ERROR: " + mess);
}

function loadImageByName(subFolder, fileName)
{
    let filePath = subFolder + "/" + fileName;
    let image_data_base64 = load_image_base64_from_path(filePath);
    return image_data_base64;
}

function load_image_base64_from_path(filePath)
{
    let image_data_base64 = fs.readFileSync(filePath).toString('base64');
    return image_data_base64;
}

function is_exists(file_path)
{
    if (fs.existsSync(file_path))
    {
        prt_mess("file " + file_path + " exists")
        return true;
    }
    else
    {
        prt_mess("file " + file_path + " doesn't exist")
        return false;
    }
}

module.exports = {
    barcode, prt_mess, is_exists, prt_err,loadImageByName, set_license
};
